data_discovery_ai/utils/es_connector.py

```diff
@@ -26,49 +26,98 @@
 def connect_es(config: configparser.ConfigParser) -> Elasticsearch:
     end_point = config["elasticsearch"]["end_point"]
     api_key = config["elasticsearch"]["api_key"]
-
-    client = Elasticsearch(end_point, api_key=api_key)
-    logging.info("Connected to ElasticSearch")
-    return client
+    try:
+        client = Elasticsearch(end_point, api_key=api_key)
+        logging.info("Connected to ElasticSearch")
+        return client
+    except Exception as e:
+        logger.error(e)
 
 
 """
     Search elasticsearch index, convert the json format to dataframe, save the dataframe to a pickle file
     Input:
         client: Elasticsearch. The initialised Elasticsearch client instance
+        index: str. The index name in Elasticsearch. Default as "es-indexer-staging"
+        batch_size: int. The number of documents searches at one query. Please not set a very large number in case to search too deep.
+        sleep_time: int. The number of seconds of sleep time between each query. This should be set with consideration of 'batch_size'. If 'batch_size' is large, set 'sleep_time' to a large number to ensure each query is finished and does not impact the next query.
     Output:
         raw_data: pd.DataFrame. The fetched raw data in a tabular format.
 """
 
 
-def search_es(client: Elasticsearch):
-    index = "es-indexer-staging"
-    start_index = 0
+def search_es(
+    client: Elasticsearch,
+    index: str = "es-indexer-staging",
+    batch_size: int = 100,
+    sleep_time: int = 5,
+):
     dataframes = []
-    batch_size = 100
-    resp = client.search(index=index, body={"query": {"match_all": {}}})
-    total_number = resp["hits"]["total"]["value"]
 
+    # get document count
+    count_resp = client.count(index=index)
+    total_number = count_resp["count"]
     rounds = (total_number + batch_size - 1) // batch_size
 
-    # TODO: refactor the search query method.
-    for round in tqdm(range(rounds), desc="searching elasticsearch"):
-        search_query = {
+    # get the first pit
+    first_pit_resp = client.open_point_in_time(
+        index=index,
+        keep_alive="1m",
+    )
+    first_pit = first_pit_resp["id"]
+
+    # the first search
+    first_query_body = {
+        "size": batch_size,
+        "query": {"match_all": {}},
+        "pit": {"id": first_pit, "keep_alive": "1m"},
+        "sort": [{"id.keyword": "asc"}],
+    }
+    first_query_resp = client.search(body=first_query_body)
+
+    data = first_query_resp["hits"]["hits"]
+    # set search after value
+    current_last_result = data[-1]["sort"]
```

Comment:
same here, might need probably try-except or validation steps to ensure you get the right inputs for further steps.